server.js

```
@@ -112,19 +137,60 @@ app.get('/dummy.js', function (req, res) {
-app.get('/parse/ahopeinit', function (req, res) {
+app.post(PARSE_MOUNT_PATH + '/ahopeinit', function (req, res) {
   console.log("Request to initialize has been received.");
-  res.type('text/plain').send("Schema initialization will start.");
+
+  // If we are running in production (i.e., process.env.MK_SALT is set), we want to
+  // prevent any unauthorized actor to request a schema init.
+  // We require an authorization header be present, with Basic as the authentication scheme,
+  // and the password part being the SHA2 hash to MK_SALT (the GCP cloud builder generates
+  // it and therefore is in possession of it).
+  //
+  if (process.env.MK_SALT) {    
+    let authenticated = false ;
+
+    if (req.headers.authorization && req.headers.authorization.toLowerCase().startsWith('basic ')) {
+      const authString = new Buffer(req.headers.authorization.split(" ")[1], 'base64').toString();
+      console.log("Authorization header detected: " + authString);
+      const secretSplit = authString.split(":")
+      if (secretSplit.length > 1)
+      {
+        const secret = secretSplit[1];
+        if (!crypto) crypto = require('crypto');
+        const hash = crypto.createHash('sha256');
+        const expectedSecret = hash.update(process.env.MK_SALT).digest('hex');
+        authenticated = (expectedSecret === secret) ;
+      }  
+    }
+
+    if (false == authenticated) {
+      console.info("Request to initialize schema is rejected as unauthorized.");
+      res.status(401).type('text/plain').send("Unauthorized.");
+      return;
+    }
+  }
+
+  // res.type('text/plain').send("Schema initialization will start.");
   const setParseSchema = require('./parseSchema');
-  setParseSchema(serverConfig);
+  const parseLocalUrl = "http://localhost:" + PORT + PARSE_MOUNT_PATH;
+  setParseSchema(serverConfig, parseLocalUrl)
+  .then(() => {
+    res.type('text/plain').send("Schema initialization completed.");
+  })
+  .catch(() => {
```

is it worth capturing the error message and logging it to the client here? this catch swallows the error message, could make debugging a little trickier if something goes wrong here

---

okay, any error should have been logged here:
https://github.com/evandana/keep-ahope-gcp/pull/8/files#diff-856c9942d7d8b144c262b3cf742d32edR244

Notice that that catch returns a rejected promise so this catch is executed and return a simple text to the REST client.